rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,775 @@
+                self.output_units = sequence_units
+            else:
+                self.return_feature_type = SENTENCE
+                self.output_units = sentence_units
+
+    def call(self, inputs: Tuple[tf.Tensor, tf.Tensor, tf.Tensor]) -> tf.Tensor:
+        """Concatenate sequence- and sentence-level feature tensors into one tensor.
+
+        Arguments:
+            inputs: Tuple containing three dense 3-D tensors.
+
+        Returns:
+            Single dense 3-D tensor containing the concatenated sequence- and sentence-
+            level features.
+        """
+        sequence = inputs[0]
+        sentence = inputs[1]
+        mask = inputs[2]
+
+        if self.return_feature_type == f"{SEQUENCE}_{SENTENCE}":
+            # If the layers for unifying the dimensions exist, apply them. They always
+            # exist both, or neither.
+            if f"unify_dims.{SEQUENCE}" in self._tf_layers:
+                sequence = self._tf_layers[f"unify_dims.{SEQUENCE}"](sequence)
+                sentence = self._tf_layers[f"unify_dims.{SENTENCE}"](sentence)
+
+            # mask has for each input example a sequence of 1s of length seq_length+1,
+            # where seq_length is the number of real tokens. The rest is 0s which form
+            # a padding up to the max. sequence length + 1 (max. # of real tokens + 1).
+            # Here the mask is turned into a mask that has 0s everywhere and 1 only at
+            # the immediate next position after the last real token's position for given
+            # input example. Example (batch size 2, sequence lengths [1, 2]):
+            # [[[1], [0], [0]],     ___\   [[[0], [1], [0]],
+            #  [[1], [1], [0]]]        /    [[0], [0], [1]]]
+            last = mask * tf.math.cumprod(
+                1 - mask, axis=1, exclusive=True, reverse=True
+            )
+
+            # The new mask is used to distribute the sentence features at the sequence
+            # positions marked by 1s. The sentence features' dimensionality effectively
+            # changes from `(batch_size, 1, feature_dim)` to `(batch_size, max_seq_length+1,
+            # feature_dim)`, but the array is sparse, with real features present only at
+            # positions determined by 1s in the mask.
+            sentence = last * sentence
+
+            # Padding of sequence-level features is increased by 1 in the sequence
+            # dimension to match the shape of modified sentence-level features.
+            sequence = tf.pad(sequence, [[0, 0], [0, 1], [0, 0]])
+
+            # Sequence- and sentence-level features effectively get concatenated by
+            # summing the two padded feature arrays like this (batch size  = 1):
+            # [[seq1, seq2, seq3, 0, 0]] + [[0, 0, 0, sent1, 0]] =
+            # = [[seq1, seq2, seq3, sent1, 0]]
+            return sequence + sentence
+
+        elif self.return_feature_type == SEQUENCE:
+            return sequence
+
+        else:
+            return sentence
```

Comment:
I think rasa convention is not to use `else` in case of return
```suggestion
        if self.return_feature_type == SEQUENCE:
            return sequence

        return sentence
```

Reply:
Hmmm, good to know 🤔